File: Protocol/_theoreme_sim/src/verification/oscillationDetector.js
/* Purpose: Detect parent flapping and report oscillation metrics over a rolling round window. */

import { THEOREM_ROOT_ID } from "../core/constants.js";

const HISTORY_WINDOW = 12;

/**
 * @param {any} state
 * @returns {{changedParents:number,flappingNodes:number,totalTracked:number,maxFlips:number}}
 */
export function updateOscillationReport(state) {
  if (!state.parentTrace) {
    state.parentTrace = new Map();
  }

  let changedParents = 0;
  let flappingNodes = 0;
  let maxFlips = 0;
  let totalTracked = 0;

  for (const [nodeId, parentId] of state.parentMap.entries()) {
    if (nodeId === THEOREM_ROOT_ID) {
      continue;
    }

    totalTracked += 1;

    const trace = state.parentTrace.get(nodeId) ?? {
      history: [],
      flips: 0,
      lastParent: null,
    };

    if (trace.lastParent !== null && trace.lastParent !== parentId) {
      trace.flips += 1;
      changedParents += 1;
    }

    trace.lastParent = parentId;
    trace.history.push(parentId);
    if (trace.history.length > HISTORY_WINDOW) {
      trace.history.shift();
    }

    const unique = new Set(trace.history.filter((item) => item !== null));
    const localFlips = countFlips(trace.history);
    if (unique.size >= 2 && localFlips >= 3) {
      flappingNodes += 1;
    }

    maxFlips = Math.max(maxFlips, trace.flips);
    state.parentTrace.set(nodeId, trace);
  }

  return {
    changedParents,
    flappingNodes,
    totalTracked,
    maxFlips,
  };
}

/**
 * @param {Array<number|null>} history
 * @returns {number}
 */
function countFlips(history) {
  let flips = 0;

  for (let i = 1; i < history.length; i += 1) {
    const prev = history[i - 1];
    const next = history[i];

    if (prev !== null && next !== null && prev !== next) {
      flips += 1;
    }
  }

  return flips;
}

File: Protocol/_theoreme_sim_py/src/verification/oscillation_detector.py
"""Purpose: Track parent flapping metrics over rolling history window."""

from __future__ import annotations

from src.core.constants import THEOREM_ROOT_ID

HISTORY_WINDOW = 12


def _count_flips(history: list[int | None]) -> int:
    flips = 0
    for index in range(1, len(history)):
        prev = history[index - 1]
        nxt = history[index]
        if prev is not None and nxt is not None and prev != nxt:
            flips += 1
    return flips


def update_oscillation_report(state: dict) -> dict:
    if not isinstance(state.get("parentTrace"), dict):
        state["parentTrace"] = {}

    changed_parents = 0
    flapping_nodes = 0
    max_flips = 0
    total_tracked = 0

    for node_id, parent_id in state["parentMap"].items():
        if node_id == THEOREM_ROOT_ID:
            continue

        total_tracked += 1

        trace = state["parentTrace"].get(
            node_id,
            {
                "history": [],
                "flips": 0,
                "lastParent": None,
            },
        )

        if trace["lastParent"] is not None and trace["lastParent"] != parent_id:
            trace["flips"] += 1
            changed_parents += 1

        trace["lastParent"] = parent_id
        trace["history"].append(parent_id)

        if len(trace["history"]) > HISTORY_WINDOW:
            trace["history"].pop(0)

        unique = {item for item in trace["history"] if item is not None}
        local_flips = _count_flips(trace["history"])
        if len(unique) >= 2 and local_flips >= 3:
            flapping_nodes += 1

        max_flips = max(max_flips, int(trace["flips"]))
        state["parentTrace"][node_id] = trace

    return {
        "changedParents": changed_parents,
        "flappingNodes": flapping_nodes,
        "totalTracked": total_tracked,
        "maxFlips": max_flips,
    }
